Ngoc_Bao/myweb/app/Http/Controllers/Admin/ProductController.php

```diff
@@ -3,84 +3,215 @@
 namespace App\Http\Controllers\Admin;
 
 use App\Http\Controllers\Controller;
+use App\Http\Controllers\Auth;
 use Illuminate\Http\Request;
 use App\Models\Product;
 use App\Models\Category;
+use App\Models\ProductHistory;
+use App\Models\ProductHistoryDetail;
 
 class ProductController extends Controller
 {
-    protected $productmodel;
-    protected $categorymodel;
+    protected $modelProduct;
+    protected $modelCategory;
+    protected $modelProductHistory;
+    protected $modelProductHistoryDetail;
 
-    public function __construct(Product $product, Category $category)
+    public function __construct(Product $product, Category $category, ProductHistory $producthistory, ProductHistoryDetail $producthistorydetail)
```

Comment:
```suggestion
    public function __construct(
        Product $product,
        Category $category,
        ProductHistory $productHistory,
        ProductHistoryDetail $productHistoryDetail
    )
```
có thể viết lại thế này cho đẹp